internal/kafka/internal/services/quota/quota_management_list_service.go

```diff
@@ -72,11 +92,19 @@ func (q QuotaManagementListService) ReserveQuota(kafka *dbapi.KafkaRequest, inst
 
 	var kafkas []*dbapi.KafkaRequest
 
+	billingModelID, err := q.detectBillingModel(kafka)
+	if err != nil {
+		return "", err
+	}
+	// TODO: find a better place to set this instead of this side effect
+	kafka.DesiredKafkaBillingModel = billingModelID
+
 	dbConn := q.connectionFactory.New().
 		Model(&dbapi.KafkaRequest{}).
-		Where("instance_type = ?", instanceType.String())
+		Where("instance_type = ?", kafka.InstanceType).
+		Where("actual_kafka_billing_model = ?", kafka.DesiredKafkaBillingModel)
```

Comment:
Could you explain further why this condition has been added and what would be the difference when not specifying it?

Reply:
In the quota_list configuration we specify how many instances we want to allow for a specific `instance_type/pair`.
If I don't specify the `actual_kafka_billing_model` I will sound all the created instances without considering the billing_model.

I'm now thinking if I shouldn't change it to:
```go
Where("actual_kafka_billing_model = ? or desired_billing_model = ?", kafka.DesiredKafkaBillingModel, kafka.DesiredKafkaBillingModel)
```

to count the already accepted requests too.

Reply:
This is a tricky question because of potential race conditions.

* If we only consider desired:
  *  It might happen that we count as being in use a kafka that ends up failing and therefore counting more than what will actually be used
  * It might happen that we don't count a kafka instance that is using that billing model (actual) but is in the process of being promoted to another type
* If we only consider actual:
  * Depending on the order of events we will not detect a kafka instance that had desired and no actual but it was in the process of setting actual. Ending up with more used kafka instances than the ones allowed
  * It might happen that we count as being in use a kafka that ends up being promoted to another type, counting more than what will actually be used
* If we consider both desired an actual then the previously explained issues of "counting more than what will end up being used" might happen

I think the approach of using an "or" is reasonable. It will potentially give us "false positives" where we will count more than what will end up being used but opting for the side of caution is what I would go for.

Reply:
Added the or clause